### PR TITLE
Task №5 ready for review

### PR DESCRIPTION
**Pull request by KseniaShurina**
There are some troubles in Task 5.2 regarding LINQ expressions

### CODE REVIEW

File: HW5/Task52/Catalog.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        public List<Book> GetBooksByAuthor(Author author)
+        {
+            var result = Books.Values.Where(b =>
+                b.Authors.Contains(author))
+                .OrderBy(b => b.PublicationDate)
+                .ToList();
+            return result;
+        }
+
+        // TODO Get from the catalog a set of tuples of the form “author’s name – the number of his books in the catalog” for all authors.
```
Comment (KseniaShurina) [inline HW5/Task52/Catalog.cs]: I have no idea how to write LINQ here

File: HW5/Task51/SparseMatrix.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        public int NumberOfRows { get; set; }
+        public int NumberOfColumns { get; set; }
+
+        private readonly Dictionary<(int, int), long> _notNullElements;
+
+        public long this[int row, int column]
+        {
+            get
+            {
```
Comment (IgorKheidorov) [inline HW5/Task51/SparseMatrix.cs:16]: move to private method

File: HW5/Task51/SparseMatrix.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                for (int i = 0; i < NumberOfRows; i++)
+                {
+                    if (this[i, j] != default)
+                    {
+                        yield return (i, j, this[i, j]);
+                    }
+                }
+            }
+        }
+
```
Comment (IgorKheidorov) [inline HW5/Task51/SparseMatrix.cs:113]: Ternary + LINQ

File: HW5/Task51/SparseMatrix.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            {
+                // Column
+                for (int j = 0; j < NumberOfColumns; j++)
+                {
+                    yield return this[i, j];
+                }
+            }
+        }
+
+        // Method returns non-zero elements with itself indexes ordered by column
```
Comment (IgorKheidorov) [inline HW5/Task51/SparseMatrix.cs:105]: LINQ

File: HW5/Task51/SparseMatrix.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                return _notNullElements[(row, column)];
+            }
+            set
+            {
+                if (row < 0 || column < 0 || row >= NumberOfRows || column >= NumberOfColumns)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                // If the value with these coordinates already exists change the value
```
Comment (IgorKheidorov) [inline HW5/Task51/SparseMatrix.cs]: check for 0

File: HW5/Task51/SparseMatrix.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                    _notNullElements.Remove((row, column));
+                    _notNullElements.Add((row, column), value);
+                }
+                _notNullElements[(row, column)] = value;
+            }
+        }
+
+        // Constructor
+        public SparseMatrix(int numberOfRows, int cumberOfColumns)
+        {
```
Comment (IgorKheidorov) [inline HW5/Task51/SparseMatrix.cs:51]: multiline

File: HW5/Task52/Entities/Book.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                }
+
+                _title = value;
+            }
+        }
+
+        public Isbn Isbn { get; set; } = null!;
+        public DateTime? PublicationDate { get; set; }
+
+        // A collection of non-repeating strings, possibly empty
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Book.cs]: HashSet<string>,use constructor , remove Add, delete methods

File: HW5/Task52/Entities/Isbn.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+{
+    public class Isbn
+    {
+        private string _isbn;
+
+        public string IsbnNumber
+        {
+            get => _isbn;
+            set
+            {
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Isbn.cs:18]: code style

File: HW5/Task52/Entities/Isbn.cs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+                if (EntityValidator.AcceptIsbn(this)) throw new ArgumentException("ISBN is not valid");
+                _isbn = value;
+            }
+        }
+
+        public Isbn(string isbn)
+        {
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Isbn.cs:30]: Regexp to const

File: HW5/Task52/Entities/Catalog.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            return result;
+        }
+
+        // Retrieve from the catalog a set of books by the specified author name, sorted alphabetically.
+        public List<Book> GetBooksByAuthor(Author author)
+        {
+            var books = Books.Values.Where(b =>
+                b.Authors.Contains(author))
+                .OrderBy(b => b.PublicationDate)
+                .ToList();
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Catalog.cs:34]: Remove

File: HW5/Task52/Entities/Catalog.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+            ArgumentNullException.ThrowIfNull(books);
+
+            return books;
+        }
+
+        // TODO Get from the catalog a set of tuples of the form “author’s name – the number of his books in the catalog” for all authors.
+        public List<(Author, int)> GetAllAuthors()
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Catalog.cs]: SelectMany
GroupBy
Select

File: HW5/Task52/Entities/Catalog.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+        //TODO Method for future tasks
+        public Book GetBookByIsbn(string isbn)
+        {
+            var convertedIsbn = IsbnConverter.ConvertIsbnToCommonPattern(isbn);
+
+            var book = Books
+                .Where(b => b.Key.ToString() == convertedIsbn)
+                .Select(b => b.Value)
+                .FirstOrDefault();
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Catalog.cs:58]: Remove

File: HW5/Task52/Entities/Isbn.cs [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            if (!Regex.IsMatch(isbn, @"\d{3}-?\d{1}-?\d{2}-?\d{6}") && !Regex.IsMatch(isbn, @"^\d{12}$"))
```
Comment (IgorKheidorov) [inline HW5/Task52/Entities/Isbn.cs]: IsbnNumber = isbn;

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please fix

**Review by IgorKheidorov — APPROVED** [review]
Ok)

**Pull request merged**